content/en/blog/2023/php-auto-instrumentation/index.md

```diff
@@ -0,0 +1,166 @@
+---
+title: Opentelemetry PHP Auto-Instrumentation
+linkTitle: PHP Auto-Instrumentation
+date: 2023-03-15
+author: '[Przemek Delewski](https://github.com/pdelewski/) (Sumo Logic)'
+---
+
+Automatic Instrumentation is a process of adding tracing capabilities into user
+application without modyfing it's source code. There are several techniques to
+do that, but all of them more or less work in the same way by injecting
+additional code into original one during compile time, link time, run-time or by
+extending the operating system in case of [ebpf](https://ebpf.io/). This
+blogpost presents method used by Opentelemetry PHP auto-instrumentation.
+
+## Preconditions
+
+Due to the technique chosen by the opentelemetry-php community for
+auto-instrumentation purposes, there are a few preconditions that must be
+fulfilled before user can use it in their application. PHP auto-instrumentation
+is based on observability API that was introduced in PHP 8.0. This means that
+user has to have at least PHP 8 installed. Another important component is the
+[composer](https://getcomposer.org/download/) package manager in most recent
+version. Third, is availability of c compiler in order to build
+[c extension](https://github.com/open-telemetry/opentelemetry-php-instrumentation)
+which is core component used for injecting tracing code.
+
+## Background
+
+[Observability API](https://www.datadoghq.com/blog/engineering/php-8-observability-baked-right-in/)
+allows register and execute additional code (function) before and after original
+one without introducing additional performance penalties in other areas, so in
+other words, we pay only for what we use and only for altered function. Before
+PHP 8, the most common technique for adding tracing capabilities was altering
+`zend_execute_ex` function (a monkey patching kind technique), however this
+could lead to stack overflow and performance problems as whole application paid
+for that. There were also considered other approaches like compile time AST
+modifications which seems feasible, howerer there are no known production ready
+tracers that works in this way up to date.
```

Comment:
```suggestion
### Background on the PHP 8.0 Observability API

[Observability API](https://www.datadoghq.com/blog/engineering/php-8-observability-baked-right-in/)
allows you to register and execute additional code (function) before and after an original
one without introducing additional performance penalties in other areas. Before
PHP 8.0, the most common technique for adding tracing capabilities was altering
the `zend_execute_ex` function (a monkey patching kind technique). However, the
use of monkey patching can lead to performance problems, stack overflows at runtime,
and a general application overhead that may not be desirable. Another approach we considered
was plugging into the AST and modifying compiled code, but there were no known-good way
to do this that was production ready.
```

Reply:
@cartermp  Last sentence changes meaning IMO. I changed a bit what you suggested.